### DISCUSSION TITLE
Regular Text file - Single column background color (shading).

### DISCUSSION THREAD

**Discussion by JontyFrancis** [shriprem/FWDataViz#53] [Q&A] [answered]
Hi, 
I have regular text files that I want to check and edit line by line.
I would like a single column of background color (shading) only for the first 19 characters wide.
This is because I'm visually breaking down each line into 19 characters or less.
I'm pretty sure FWDataViz can do it but I'm lost as to how to configure it.
Please can you tell me how to do it?
Much thanks!
John

**Comment by shriprem** [comment, marked as answer]
@JontyFrancis, here is the basic file type definition based on your specified requirements.

### File-type Definition to Import
```INI
[FT009_FIRST_19_CHARS]
FileLabel=First 19 Chars
FileTheme=Spectrum
RecordTerminator=
MultiByteChars=N
ADFT_Line_01=
ADFT_Regex_01=
ADFT_Line_02=
ADFT_Regex_02=
ADFT_Line_03=
ADFT_Regex_03=
RecordTypes=REC001
REC001_Label=First 19
REC001_Marker=
REC001_FieldLabels=first_19,later_chars
REC001_FieldWidths=19,3000
```

###  How to Import
1. In the _FWDataViz_ panel, click on the **File Type Metadata Editor** button.
![image](https://user-images.githubusercontent.com/45252729/169658395-1486aa18-8bbb-415f-8e5d-bcbf1b3eca23.png)

2. In the _File Type Metadata Editor_, click on the **Append New File Type Data** button.
3. In the _Append File Type Metadata_ window, paste the file type definition data provided above.
4. Click the **Append File Type** button to close the _Append File Type Metadata_ window.
5. Back in the _File Type Metadata Editor_ window, click on the **Save as Primary Configuration** button.

### Notes
1. Since I do not know much about the files you are working with, I did not include an _Automatic File Detection_ pattern for the file type. So, each time you need _FWDataViz_ to color the first 19 characters, you will need to manually select this new file type in the dropdown list in the _FWDataViz_ panel.
2. You may also be aware that within Notepad++ itself, you can use the _Preferences_ » _Margins/Border/Edge_ section » _Vertical Edge Settings_  to set a vertical edge at 19 characters. Although, with _FWDataViz_, you also get the option of coloring either with character counts or byte counts. The byte counts assume importance if you have TAB characters or Unicode characters in the text.

**Reply by JontyFrancis** [reply]
Thanks Shridhar, 

Thanks for your reply! I think I didn't explain it too well. 

I wanted a block of shading for the whole file - covering only the first 19 characters/positions of each line.
The shading should be there in the background (1-19) basically ignoring whatever characters are (or are not) in the line. 

The shading starts at position 1 on each line - for 19 positions - for all the lines in the file.

If there are less than 19 characters on a line, the shading should still be 19 character positions wide (if possible).

Having said that...

Your Note #2 gives me the correct effect I want - 
but it highlights the characters 19+ (over 19) rather than 1->19. 

That's a good work around - it helps a lot. 

So thanks very much for that suggestion!

Still, if it's possible to do what I want with FWDataViz as outlined above - that would be great too.

I really appreciate your help!

Thanks again! 

John

**Comment by shriprem** [comment]
I assume this is what you are looking for:
1. The first 19 characters are shaded with just a background color. Text color remains black (i.e., the default global text color in Light mode in NPP). And the text is neither bolded nor in italics.
2. The characters beyond the 19th remain _unstyled_ in the default global background and text color.

I also assume that the file is not being styled by any language-specific lexer like HTML, CSS, JavaScript, SQL, etc. If it is being styled by a language-specific lexer, then the styling by _FWDataViz_ will be in conflict with it.

This can be accomplished with a custom Theme definition with the output like this:
![image](https://user-images.githubusercontent.com/45252729/169672210-85fb39a1-a031-4887-8ea0-725233dbd8c8.png)

If you confirm this is what you want, I will post the custom Theme definition. You can then associate that Theme with the File Type definition I gave you earlier. But first please confirm. Thanks.

**Reply by JontyFrancis** [reply]
That's exactly what I want - all your assumptions are correct.
Thanks very much!

**Comment by shriprem** [comment]
But I also noticed this line later:
> If there are less than 19 characters on a line, the shading should still be 19 character positions wide (if possible).

Shading up to the 19th character position even if a line has fewer than 19 characters is not possible with Scintilla, the engine behind NPP. The Scintilla styles will apply only when there are characters (or even spaces or some other control characters) at each position up to the 19th on every line. You will notice this even in NPP, when you set the _Background Mode_ for the _Vertical Edge_. 

So with this disclaimer, here are the instructions for the custom theme.

1. Click the Visualizer Theme button in the FWDataViz side panel.
![image](https://user-images.githubusercontent.com/45252729/169672920-ed3d6fa6-0f4a-48d9-a2d5-f9344ebb4353.png)

2.  Click the **Append New Theme Data** button in the Theme Editor window.
3. In the _Append Theme Data_ window, paste this theme definition:
```ini
[First_19_Only]
Count=2
EOL=ECFFFF 000000 00
BFBI_00=4FE9FC 000000 00
BFBI_01=ECFFFF 000000 00
```
4. Click the **Append File Type** button. The _Append Theme Data_ window will close.
5.  Click the **Save as Primary Themes File** button in the Theme Editor.
6. In the _FWDataViz_ panel, click on the **File Type Metadata Editor** button.
7. Select the _First 19 Chars_ file type, and associate it with the theme we just added. Like so:
![image](https://user-images.githubusercontent.com/45252729/169673101-60d378c1-bedf-4e89-9a95-868c216a52a6.png)
8. Click on the **Save as Primary Configuration** button.

If you review the custom file type and theme definitions in their respective editors provided by the _FWDataViz_ plugin, you will quickly figure out how they work together. You can then tweak it further to suit your color preferences.

**Comment by JontyFrancis** [comment]
Perfect Shridhar! 
Thank you so much!
